library/core/src/num/mod.rs

```diff
@@ -278,18 +277,12 @@ macro_rules! widening_impl {
 
 macro_rules! conv_rhs_for_unchecked_shift {
     ($SelfT:ty, $x:expr) => {{
-        #[inline]
-        fn conv(x: u32) -> $SelfT {
-            // FIXME(const-hack) replace with `.try_into().ok().unwrap_unchecked()`.
-            // SAFETY: Any legal shift amount must be losslessly representable in the self type.
-            unsafe { x.try_into().ok().unwrap_unchecked() }
-        }
-        #[inline]
-        const fn const_conv(x: u32) -> $SelfT {
-            x as _
+        // If the `as` cast will truncate, ensure we still tell the backend
+        // that the pre-truncation value was also small.
+        if <$SelfT>::BITS < 32 {
```

Comment:
Why not assume unconditionally?

Reply:
Because it's superfluous.

The `unchecked_sh[lr]` intrinsics are already UB for a shift that's out of range, they lower to <https://llvm.org/docs/LangRef.html#shl-instruction> that are ~~UB~~poison for an out-of-range shift, and CTFE checks for the UB <https://github.com/rust-lang/rust/blob/6bba061467f7c2cab04b262b95eb67bf89265587/compiler/rustc_const_eval/src/interpret/intrinsics.rs#L256-L264>.

So if it's worth assuming that, I think that should be the job of the codegen backend when it emits `intrinsics::unchecked_shl`, not something that we ought to carry around in MIR.

(It's just to keep from losing the UB in `3_u16.unchecked_shl(0x10001_u32)` as part of calling the intrinsic.)

Reply:
I agree that `assume` is superfluous in those cases. OTOH, a branch to avoid that `assume` makes analyses more complex, and makes IR heavier.

Reply:
> OTOH, a branch to avoid that `assume` makes analyses more complex, and makes IR heavier.

Check the MIR, though -- it's a constant condition, so the branch is always optimized out.

(Fully agreed that a conditional assume would not be worth it.)